File: web/src/flyover/flyover-path.ts
/**
 * Pure geometry for the hole flyover: waypoint smoothing (Catmull-Rom),
 * arc-length lookup along the smoothed path, the low-flight camera pose for
 * a given path position, and the speed profile (constant 225 km/h with an
 * eased start and stop). No MapLibre, no DI — everything here is unit-tested
 * in `tests/flyover-path.test.ts`.
 *
 * Distances are metres on a local equirectangular plane anchored at the
 * path's first waypoint; a hole is < 1 km so the projection error is
 * negligible for camera placement.
 */

export interface LngLat {
    lng: number;
    lat: number;
}

/** A point on the resampled path with its cumulative arc length `s` (m). */
export interface PathPoint extends LngLat {
    s: number;
}

export interface FlyoverPath {
    points: PathPoint[];
    /** Total arc length in metres. */
    length: number;
}

/** Camera placement expressed as "from here, look at there" in WGS84 + metres. */
export interface CameraPose {
    from: LngLat;
    /** Camera altitude, metres (already multiplied by terrain exaggeration). */
    altFrom: number;
    to: LngLat;
    /** Target altitude, metres (already multiplied by terrain exaggeration). */
    altTo: number;
    /** Compass bearing from camera to target, degrees [0, 360). */
    bearing: number;
    /** MapLibre pitch: 0 = straight down, 90 = horizon. */
    pitch: number;
}

export interface PoseOptions {
    /** Eye height above local ground, metres (before exaggeration). */
    eyeHeightM: number;
    /** The eye looks at the path point this far ahead (clamped to the path end). */
    lookAheadM: number;
    /** The look point sits this high above the ground there, metres. */
    lookHeightM: number;
    /** Terrain vertical exaggeration; scales ground + heights. */
    exaggeration: number;
}

/** Eye height above local ground while flying, metres. */
export const FLYOVER_EYE_HEIGHT_M = 5;
/** Look target: this far ahead along the path, clamped to the green. */
export const FLYOVER_LOOK_AHEAD_M = 120;
/** Look target height above the ground at that point, metres. */
export const FLYOVER_LOOK_HEIGHT_M = 2;
/**
 * The eye stops this short of the green centre. Without it the final frame
 * would have the eye above the green looking straight down at its own look
 * point (from == to); 15 m back at 5 m up gives an approach view of the green.
 */
export const FLYOVER_END_STANDOFF_M = 15;
/** Cruise speed along the path: 225 km/h. */
export const FLYOVER_SPEED_M_S = 225 / 3.6;
/** Ease-in at the tee and ease-out at the green, each this long, ms. */
export const FLYOVER_EASE_MS = 1500;
/** Shortest flight (very short holes), ms. The only duration clamp. */
export const FLYOVER_MIN_FLIGHT_MS = 3000;
/** Overhead camera → start pose transition, ms (same as walk mode's enter). */
export const FLYOVER_ENTER_MS = 1500;
/** Hold at the final pose before restoring the overhead camera, ms. */
export const FLYOVER_HOLD_MS = 2000;
/** Resampling step for the smoothed path, metres. */
export const PATH_STEP_M = 5;

const EARTH_RADIUS_M = 6371008.8;
const DEG = Math.PI / 180;

/** Metres per degree of longitude / latitude at a reference latitude. */
export function metersPerDegree(latDeg: number): { x: number; y: number } {
    const perDegLat = EARTH_RADIUS_M * DEG;
    return { x: perDegLat * Math.cos(latDeg * DEG), y: perDegLat };
}

/** Planar distance in metres (equirectangular, fine for sub-km spans). */
export function distanceM(a: LngLat, b: LngLat): number {
    const m = metersPerDegree((a.lat + b.lat) / 2);
    const dx = (b.lng - a.lng) * m.x;
    const dy = (b.lat - a.lat) * m.y;
    return Math.hypot(dx, dy);
}

/** Compass bearing a→b in degrees [0, 360). */
export function bearingDeg(a: LngLat, b: LngLat): number {
    const m = metersPerDegree((a.lat + b.lat) / 2);
    const dx = (b.lng - a.lng) * m.x;
    const dy = (b.lat - a.lat) * m.y;
    const deg = Math.atan2(dx, dy) / DEG;
    return ((deg % 360) + 360) % 360;
}

/**
 * Pick the back tee: the tee farthest from the green centre. Null for an
 * empty list. Ties resolve to the first in list order.
 */
export function backTee<T extends { lat: number; lon: number }>(
    tees: readonly T[],
    green: { lat: number; lon: number },
): T | null {
    let best: T | null = null;
    let bestD = -1;
    const g = { lng: green.lon, lat: green.lat };
    for (const tee of tees) {
        const d = distanceM({ lng: tee.lon, lat: tee.lat }, g);
        if (d > bestD) {
            best = tee;
            bestD = d;
        }
    }
    return best;
}

function catmullRom(p0: number, p1: number, p2: number, p3: number, t: number): number {
    const t2 = t * t;
    const t3 = t2 * t;
    return 0.5 * (
        2 * p1
        + (-p0 + p2) * t
        + (2 * p0 - 5 * p1 + 4 * p2 - p3) * t2
        + (-p0 + 3 * p1 - 3 * p2 + p3) * t3
    );
}

/**
 * Smooth the waypoints with a Catmull-Rom spline (end tangents from
 * duplicated endpoints) and resample it at roughly `stepM` metres. Two
 * waypoints give a straight line. Consecutive duplicates (< 0.5 m apart)
 * collapse. Returns null when fewer than two distinct waypoints remain.
 */
export function buildFlyoverPath(waypoints: readonly LngLat[], stepM = PATH_STEP_M): FlyoverPath | null {
    const distinct: LngLat[] = [];
    for (const w of waypoints) {
        const prev = distinct[distinct.length - 1];
        if (!prev || distanceM(prev, w) >= 0.5) distinct.push({ lng: w.lng, lat: w.lat });
    }
    if (distinct.length < 2) return null;

    const origin = distinct[0];
    const m = metersPerDegree(origin.lat);
    const toPlane = (p: LngLat) => ({ x: (p.lng - origin.lng) * m.x, y: (p.lat - origin.lat) * m.y });
    const toLngLat = (x: number, y: number): LngLat => ({ lng: origin.lng + x / m.x, lat: origin.lat + y / m.y });
    const plane = distinct.map(toPlane);

    // Sample the spline densely (chord-proportional), then accumulate s.
    const samples: Array<{ x: number; y: number }> = [plane[0]];
    for (let i = 0; i < plane.length - 1; i++) {
        const p0 = plane[Math.max(0, i - 1)];
        const p1 = plane[i];
        const p2 = plane[i + 1];
        const p3 = plane[Math.min(plane.length - 1, i + 2)];
        const chord = Math.hypot(p2.x - p1.x, p2.y - p1.y);
        const n = Math.max(1, Math.ceil(chord / stepM));
        for (let k = 1; k <= n; k++) {
            const t = k / n;
            if (plane.length === 2) {
                samples.push({ x: p1.x + (p2.x - p1.x) * t, y: p1.y + (p2.y - p1.y) * t });
            } else {
                samples.push({
                    x: catmullRom(p0.x, p1.x, p2.x, p3.x, t),
                    y: catmullRom(p0.y, p1.y, p2.y, p3.y, t),
                });
            }
        }
    }

    const points: PathPoint[] = [];
    let s = 0;
    for (let i = 0; i < samples.length; i++) {
        if (i > 0) s += Math.hypot(samples[i].x - samples[i - 1].x, samples[i].y - samples[i - 1].y);
        points.push({ ...toLngLat(samples[i].x, samples[i].y), s });
    }
    return { points, length: s };
}

/**
 * Position at arc length `s`. Values outside [0, length] extrapolate
 * linearly along the first / last segment so a camera trailing the tee or a
 * look-at point past the green still lies on the hole's line.
 */
export function pointAlong(path: FlyoverPath, s: number): LngLat {
    const pts = path.points;
    if (pts.length === 1) return { lng: pts[0].lng, lat: pts[0].lat };

    let a: PathPoint;
    let b: PathPoint;
    if (s <= 0) {
        a = pts[0];
        b = pts[1];
    } else if (s >= path.length) {
        a = pts[pts.length - 2];
        b = pts[pts.length - 1];
    } else {
        // Binary search for the segment containing s.
        let lo = 0;
        let hi = pts.length - 1;
        while (hi - lo > 1) {
            const mid = (lo + hi) >> 1;
            if (pts[mid].s <= s) lo = mid;
            else hi = mid;
        }
        a = pts[lo];
        b = pts[hi];
    }
    const span = b.s - a.s;
    const t = span > 0 ? (s - a.s) / span : 0;
    return { lng: a.lng + (b.lng - a.lng) * t, lat: a.lat + (b.lat - a.lat) * t };
}

/**
 * Ground elevation along the path, sampled at fixed arc-length steps
 * starting at `s0` (which may be negative to cover the camera's trailing
 * position). `values` are unexaggerated metres above sea level.
 */
export interface GroundProfile {
    s0: number;
    stepM: number;
    values: number[];
}

/** Arc-length positions a ground profile should be sampled at. */
export function groundSampleStations(path: FlyoverPath, behindM: number, aheadM: number, stepM = PATH_STEP_M): { s0: number; stations: number[] } {
    const s0 = -behindM - stepM;
    const end = path.length + aheadM + stepM;
    const stations: number[] = [];
    for (let s = s0; s <= end + 1e-9; s += stepM) stations.push(s);
    return { s0, stations };
}

/**
 * Build a profile from raw samples, filling gaps (null = outside tile
 * coverage) from the nearest known neighbour. All-null → flat zero.
 */
export function fillGroundProfile(s0: number, stepM: number, raw: ReadonlyArray<number | null>): GroundProfile {
    const values = new Array<number>(raw.length);
    let last: number | null = null;
    for (let i = 0; i < raw.length; i++) {
        const v = raw[i];
        if (v !== null && Number.isFinite(v)) last = v;
        values[i] = last ?? Number.NaN;
    }
    // Leading gap: back-fill from the first known value.
    let first: number | null = null;
    for (const v of values) if (!Number.isNaN(v)) { first = v; break; }
    for (let i = 0; i < values.length; i++) values[i] = Number.isNaN(values[i]) ? (first ?? 0) : values[i];
    return { s0, stepM, values };
}

/** Linearly interpolated ground height at arc length `s`, clamped at the ends. */
export function groundAt(profile: GroundProfile, s: number): number {
    const { s0, stepM, values } = profile;
    if (values.length === 0) return 0;
    const f = (s - s0) / stepM;
    if (f <= 0) return values[0];
    if (f >= values.length - 1) return values[values.length - 1];
    const i = Math.floor(f);
    const t = f - i;
    return values[i] + (values[i + 1] - values[i]) * t;
}

/** Pitch (MapLibre convention) for a camera `dAlt` above its target at `horizontalM`. */
export function pitchFor(horizontalM: number, dAlt: number): number {
    return 90 - Math.atan2(dAlt, horizontalM) / DEG;
}

/** Compose a pose from camera + target positions and altitudes. */
export function poseFromTo(from: LngLat, altFrom: number, to: LngLat, altTo: number): CameraPose {
    return {
        from,
        altFrom,
        to,
        altTo,
        bearing: bearingDeg(from, to),
        pitch: pitchFor(distanceM(from, to), altFrom - altTo),
    };
}

/** Arc length where the eye stops: `standoffM` short of the green, never below 0. */
export function eyePathLength(path: FlyoverPath, standoffM = FLYOVER_END_STANDOFF_M): number {
    return Math.max(0, path.length - standoffM);
}

/**
 * Camera pose while flying: the eye is on the path at arc length `s`,
 * `eyeHeightM` above the ground there, and looks at the path point
 * `lookAheadM` ahead (clamped to the green) at `lookHeightM` above ground.
 * Ground and heights scale by `exaggeration` to match MapLibre's exaggerated
 * terrain mesh.
 *
 * The geometric pitch is 90 - atan((eyeH - lookH) / lookAhead): with the
 * defaults 5 m / 2 m / 120 m that is 88.6°, past MapLibre's 85° maximum. The
 * pose reports the geometric value; the service raises `maxPitch` to 85 and
 * MapLibre clamps on `jumpTo` (the eye then sits ~12.5 m above ground, i.e.
 * lookAhead * tan(5°) + lookH, with the look point fixed). As the eye nears the green the look distance shrinks
 * to `FLYOVER_END_STANDOFF_M` and the pitch drops to ~79°, under the cap.
 *
 * Heading is the bearing eye → look point. Because the look point is 120 m
 * ahead on a Catmull-Rom spline, the heading is a running chord over 120 m of
 * path: corner turns spread over the approach rather than stepping at the
 * aim point (`tests/flyover-path.test.ts` bounds the per-step change).
 */
export function cameraPose(
    path: FlyoverPath,
    s: number,
    ground: (s: number) => number,
    opts: PoseOptions,
): CameraPose {
    const sTo = Math.min(path.length, s + opts.lookAheadM);
    const from = pointAlong(path, s);
    const to = pointAlong(path, sTo);
    const altFrom = (ground(s) + opts.eyeHeightM) * opts.exaggeration;
    const altTo = (ground(sTo) + opts.lookHeightM) * opts.exaggeration;
    return poseFromTo(from, altFrom, to, altTo);
}

/** Linear blend between two poses (positions + altitudes; angles recomputed). */
export function lerpPose(a: CameraPose, b: CameraPose, t: number): CameraPose {
    const mix = (x: number, y: number) => x + (y - x) * t;
    return poseFromTo(
        { lng: mix(a.from.lng, b.from.lng), lat: mix(a.from.lat, b.from.lat) },
        mix(a.altFrom, b.altFrom),
        { lng: mix(a.to.lng, b.to.lng), lat: mix(a.to.lat, b.to.lat) },
        mix(a.altTo, b.altTo),
    );
}

/** Symmetric ease in/out (cubic). */
export function easeInOutCubic(t: number): number {
    const x = Math.min(1, Math.max(0, t));
    return x < 0.5 ? 4 * x * x * x : 1 - Math.pow(-2 * x + 2, 3) / 2;
}

/**
 * Flight duration for `lengthM` of eye path at constant `speedMs`, with a
 * speed ramp of `easeMs` at each end. A symmetric ramp covers half the
 * distance a constant-speed segment would, so the two ramps together cost
 * exactly one `easeMs` over the pure `length / speed` time. Floor: `minMs`.
 */
export function flyoverDurationMs(
    lengthM: number,
    speedMs = FLYOVER_SPEED_M_S,
    easeMs = FLYOVER_EASE_MS,
    minMs = FLYOVER_MIN_FLIGHT_MS,
): number {
    return Math.max(minMs, (lengthM / speedMs) * 1000 + easeMs);
}

/** Integral of smoothstep(x) = 3x² − 2x³ from 0 to x. */
function smoothstepIntegral(x: number): number {
    return x * x * x - x * x * x * x / 2;
}

/**
 * Fraction of the path covered at normalised time `u` ∈ [0, 1] when the
 * speed ramps up with a smoothstep over the first `ramp` of the time, holds
 * constant, and ramps down over the last `ramp` (ramp ≤ 0.5). Speed is
 * continuous and so is acceleration at the ramp ends. With
 * `ramp = easeMs / durationMs` the cruise speed equals `FLYOVER_SPEED_M_S`
 * for any duration `flyoverDurationMs` returns above the floor.
 */
export function flightProgress(u: number, ramp: number): number {
    const x = Math.min(1, Math.max(0, u));
    const r = Math.min(0.5, Math.max(0, ramp));
    // Normalising area: r/2 per ramp + (1 − 2r) cruise = 1 − r.
    const area = 1 - r;
    if (r === 0) return x;
    let d: number;
    if (x < r) d = r * smoothstepIntegral(x / r);
    else if (x <= 1 - r) d = r / 2 + (x - r);
    else d = area - r * smoothstepIntegral((1 - x) / r);
    return d / area;
}

File: ios/GolfMap/Map/FlyoverPath.swift
import Foundation

/// Pure geometry for the hole flyover: the smoothed ground path (tee → aim
/// points → green centre), the camera pose along it, and the timing. No map
/// or UIKit dependency, so the whole thing is unit-testable; `FlyoverAnimator`
/// turns poses into `MLNMapCamera`s per display frame.
///
/// All lengths are metres. Path maths runs in projected SWEREF 99 TM
/// (EPSG:3006), like `Distance.planarMeters`; headings are degrees clockwise
/// from north.
public struct FlyoverPath: Equatable, Sendable {
    /// A planar point on the densified path.
    public struct Vertex: Equatable, Sendable {
        public var x: Double
        public var y: Double
        /// Arc length from the path start.
        public var distance: Double
    }

    public let vertices: [Vertex]

    /// Total arc length.
    public var length: Double { vertices.last?.distance ?? 0 }

    /// Builds a Catmull-Rom spline through `waypoints` (the spline passes
    /// through every waypoint) and samples it about every `sampleSpacing`
    /// metres. Consecutive duplicate waypoints collapse; two distinct
    /// waypoints give a straight line. Nil with fewer than two distinct points.
    public static func build(waypoints: [LatLon], sampleSpacing: Double = 5) -> FlyoverPath? {
        var control: [(x: Double, y: Double)] = []
        for waypoint in waypoints {
            let p = Sweref99TM.fromWGS84(waypoint)
            if let last = control.last, hypot(p.x - last.x, p.y - last.y) < 0.5 { continue }
            control.append((p.x, p.y))
        }
        guard control.count >= 2 else { return nil }

        // Duplicate the end points so the spline is defined over every segment.
        let padded = [control[0]] + control + [control[control.count - 1]]
        var samples: [(x: Double, y: Double)] = [control[0]]
        let spacing = max(sampleSpacing, 0.5)
        for i in 1..<(padded.count - 2) {
            let p0 = padded[i - 1], p1 = padded[i], p2 = padded[i + 1], p3 = padded[i + 2]
            let chord = hypot(p2.x - p1.x, p2.y - p1.y)
            let steps = max(1, Int((chord / spacing).rounded(.up)))
            for step in 1...steps {
                let t = Double(step) / Double(steps)
                samples.append(catmullRom(p0, p1, p2, p3, t))
            }
        }

        var vertices: [Vertex] = []
        var distance = 0.0
        for (index, s) in samples.enumerated() {
            if index > 0 {
                let prev = samples[index - 1]
                distance += hypot(s.x - prev.x, s.y - prev.y)
            }
            vertices.append(Vertex(x: s.x, y: s.y, distance: distance))
        }
        return FlyoverPath(vertices: vertices)
    }

    private static func catmullRom(
        _ p0: (x: Double, y: Double), _ p1: (x: Double, y: Double),
        _ p2: (x: Double, y: Double), _ p3: (x: Double, y: Double),
        _ t: Double
    ) -> (x: Double, y: Double) {
        let t2 = t * t, t3 = t2 * t
        func axis(_ a: Double, _ b: Double, _ c: Double, _ d: Double) -> Double {
            0.5 * ((2 * b) + (-a + c) * t + (2 * a - 5 * b + 4 * c - d) * t2 + (-a + 3 * b - 3 * c + d) * t3)
        }
        return (axis(p0.x, p1.x, p2.x, p3.x), axis(p0.y, p1.y, p2.y, p3.y))
    }

    /// Planar point at arc length `distance` (clamped to the path; the path
    /// extends as a straight line beyond either end for negative or
    /// past-the-end distances so a trailing camera eye has a position).
    public func point(at distance: Double) -> (x: Double, y: Double) {
        guard let first = vertices.first, let last = vertices.last else { return (0, 0) }
        if vertices.count == 1 { return (first.x, first.y) }
        if distance <= 0 {
            let h = heading(at: 0) * .pi / 180
            return (first.x + sin(h) * distance, first.y + cos(h) * distance)
        }
        if distance >= last.distance {
            let h = heading(at: last.distance) * .pi / 180
            let over = distance - last.distance
            return (last.x + sin(h) * over, last.y + cos(h) * over)
        }
        let index = segmentIndex(at: distance)
        let a = vertices[index], b = vertices[index + 1]
        let span = b.distance - a.distance
        let t = span > 0 ? (distance - a.distance) / span : 0
        return (a.x + (b.x - a.x) * t, a.y + (b.y - a.y) * t)
    }

    /// WGS84 point at arc length `distance`.
    public func coordinate(at distance: Double) -> LatLon {
        let p = point(at: distance)
        return Sweref99TM.toWGS84(x: p.x, y: p.y)
    }

    /// Direction of travel at arc length `distance`, degrees clockwise from
    /// north. Constant along a segment; the end segments extend past the ends.
    public func heading(at distance: Double) -> Double {
        guard vertices.count >= 2 else { return 0 }
        let clamped = min(max(distance, 0), length)
        let index = clamped >= length ? vertices.count - 2 : segmentIndex(at: clamped)
        let a = vertices[index], b = vertices[index + 1]
        return Self.headingDegrees(dx: b.x - a.x, dy: b.y - a.y)
    }

    /// Index of the segment `[i, i+1]` containing `distance` (0 ≤ distance < length).
    private func segmentIndex(at distance: Double) -> Int {
        var lo = 0, hi = vertices.count - 2
        while lo < hi {
            let mid = (lo + hi + 1) / 2
            if vertices[mid].distance <= distance { lo = mid } else { hi = mid - 1 }
        }
        return lo
    }

    /// Planar direction to a compass heading in `[0, 360)`.
    static func headingDegrees(dx: Double, dy: Double) -> Double {
        let degrees = atan2(dx, dy) * 180 / .pi
        return degrees < 0 ? degrees + 360 : degrees
    }
}

/// One camera frame of the flyover, in MapLibre camera terms.
public struct FlyoverPose: Equatable, Sendable {
    /// Ground point the camera looks at.
    public var center: LatLon
    /// Camera eye height above the map plane at `center`, metres.
    public var altitude: Double
    /// Tilt from straight down, degrees (0 = top-down).
    public var pitch: Double
    /// Compass heading the camera faces, degrees clockwise from north.
    public var heading: Double
}

/// A complete flyover: path, ground profile and the camera/timing model.
public struct FlyoverPlan: Equatable, Sendable {
    /// Camera eye height above the local ground.
    public static let eyeHeight = 45.0
    /// The look-at point runs this far ahead of the current path point.
    public static let lookAhead = 120.0
    /// The eye trails this far behind the current path point.
    public static let trailBehind = 80.0
    /// MapLibre iOS caps camera pitch at 60°; a steeper geometric pitch pulls
    /// the look-at point back towards the eye instead.
    public static let maxPitch = 60.0
    /// Ground speed: 300 m in 8 s.
    public static let speed = 300.0 / 8.0
    /// Shortest flight, so a par-3 does not flash past.
    public static let minimumFlightDuration = 4.0
    /// Speed ramps at the start and end.
    public static let rampDuration = 1.5
    /// Hold above and behind the green after the flight.
    public static let holdDuration = 2.0
    /// Spacing of the ground-elevation profile samples.
    public static let profileSpacing = 10.0

    public let path: FlyoverPath
    /// Ground elevation sampled every `profileSpacing` metres from
    /// `-trailBehind` (the eye's starting position) to the path end.
    public let groundProfile: [Double]

    public init(path: FlyoverPath, groundProfile: [Double]) {
        self.path = path
        self.groundProfile = groundProfile
    }

    /// The WGS84 points a caller samples ground elevation at, in the order
    /// `init(path:groundElevations:)` expects them.
    public static func profileSamplePositions(for path: FlyoverPath) -> [LatLon] {
        profileDistances(for: path).map { path.coordinate(at: $0) }
    }

    static func profileDistances(for path: FlyoverPath) -> [Double] {
        var distances: [Double] = []
        var d = -trailBehind
        while d < path.length {
            distances.append(d)
            d += profileSpacing
        }
        distances.append(path.length)
        return distances
    }

    /// Builds a plan from raw elevation samples (one per
    /// `profileSamplePositions`); missing samples take the nearest known
    /// value, and an all-nil profile is flat.
    public init(path: FlyoverPath, groundElevations: [Double?]) {
        var filled: [Double] = []
        let known = groundElevations.enumerated().compactMap { index, value in
            value.map { (index: index, value: $0) }
        }
        for index in groundElevations.indices {
            if let value = groundElevations[index] {
                filled.append(value)
            } else if let nearest = known.min(by: { abs($0.index - index) < abs($1.index - index) }) {
                filled.append(nearest.value)
            } else {
                filled.append(0)
            }
        }
        self.init(path: path, groundProfile: filled)
    }

    /// Ground elevation at arc length `distance`, linearly interpolated from
    /// the profile (clamped at both ends). 0 without a profile.
    public func groundElevation(at distance: Double) -> Double {
        guard !groundProfile.isEmpty else { return 0 }
        let position = (distance + Self.trailBehind) / Self.profileSpacing
        let lower = Int(position.rounded(.down))
        if lower < 0 { return groundProfile[0] }
        if lower >= groundProfile.count - 1 { return groundProfile[groundProfile.count - 1] }
        let t = position - Double(lower)
        return groundProfile[lower] + (groundProfile[lower + 1] - groundProfile[lower]) * t
    }

    // MARK: Timing

    /// Flight time (excluding the end hold).
    public var flightDuration: Double {
        max(Self.minimumFlightDuration, path.length / Self.speed)
    }

    /// Flight plus hold.
    public var totalDuration: Double { flightDuration + Self.holdDuration }

    /// Distance along the path at time `t`: a trapezoidal speed profile with
    /// `rampDuration` ramps at each end. Clamped to `[0, length]`.
    public func distance(atTime t: Double) -> Double {
        let duration = flightDuration
        let length = path.length
        if t <= 0 { return 0 }
        if t >= duration { return length }
        let ramp = min(Self.rampDuration, duration / 2)
        let peakSpeed = length / (duration - ramp)
        if t < ramp {
            return peakSpeed * t * t / (2 * ramp)
        }
        if t <= duration - ramp {
            return peakSpeed * (t - ramp / 2)
        }
        let remaining = duration - t
        return length - peakSpeed * remaining * remaining / (2 * ramp)
    }

    // MARK: Pose

    /// Camera pose at time `t` (the end pose during the hold).
    public func pose(atTime t: Double) -> FlyoverPose {
        pose(atDistance: distance(atTime: min(t, flightDuration)))
    }

    /// Camera pose for the path point at arc length `distance`: the eye trails
    /// `trailBehind` metres behind it at `eyeHeight` above its own ground,
    /// looking at the point `lookAhead` metres further along the path (capped
    /// at the path end, so the last frames look at the green from behind).
    /// The eye altitude is relative to the look-at point's ground, so a tee
    /// above the green raises the camera instead of driving it into the map.
    public func pose(atDistance distance: Double) -> FlyoverPose {
        let clamped = min(max(distance, 0), path.length)
        let current = path.point(at: clamped)
        let headingRad = path.heading(at: clamped) * .pi / 180
        let eye = (
            x: current.x - sin(headingRad) * Self.trailBehind,
            y: current.y - cos(headingRad) * Self.trailBehind
        )
        let targetDistance = min(clamped + Self.lookAhead, path.length)
        var target = path.point(at: targetDistance)

        var dx = target.x - eye.x
        var dy = target.y - eye.y
        var horizontal = hypot(dx, dy)
        if horizontal < 1 {
            dx = sin(headingRad)
            dy = cos(headingRad)
            horizontal = 1
        }
        let heading = FlyoverPath.headingDegrees(dx: dx, dy: dy)

        let groundAtEye = groundElevation(at: clamped - Self.trailBehind)
        let groundAtTarget = groundElevation(at: targetDistance)
        let altitude = max(Self.eyeHeight + (groundAtEye - groundAtTarget), Self.eyeHeight / 2)

        var pitch = atan2(horizontal, altitude) * 180 / .pi
        if pitch > Self.maxPitch {
            pitch = Self.maxPitch
            let reach = altitude * tan(Self.maxPitch * .pi / 180)
            target = (eye.x + dx / horizontal * reach, eye.y + dy / horizontal * reach)
        }

        return FlyoverPose(
            center: Sweref99TM.toWGS84(x: target.x, y: target.y),
            altitude: altitude,
            pitch: pitch,
            heading: heading
        )
    }
}

/// A flyover request for `CourseMapView`; applied once per `token` change.
public struct FlyoverCommand: Equatable, Sendable {
    public var plan: FlyoverPlan
    public var token: Int

    public init(plan: FlyoverPlan, token: Int) {
        self.plan = plan
        self.token = token
    }
}
